File: cpp/Problem90.cpp
/*
 * Problem: 90. Subsets II
 * URL: https://leetcode.com/problems/subsets-ii/
 * Difficulty: Medium
 * Tags: Array, Permutation, Backtracking
 */

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
	vector<vector<int>> subsetsWithDup(vector<int>& nums) {
		vector<vector<int>> subsets;
		sort(nums.begin(), nums.end());
		backtrack(subsets, nums, 0);
		return subsets;
	}

private:
	vector<int> permutations;
	void backtrack(vector<vector<int>> &results, vector<int> &nums, int index) {
		results.push_back(permutations);
		for (int i = index; i < nums.size(); i++) {
			if (i != index && nums[i] == nums[i - 1]) {
				continue;
			}
			permutations.push_back(nums[i]);
			backtrack(results, nums, i + 1);
			permutations.pop_back();
		}
	}
};

File: python/Problem90.py
# Problem: 90. Subsets II
# URL: https://leetcode.com/problems/subsets-ii/
# Difficulty: Medium
# Tags: Array, Permutation, Backtracking

from typing import List

class Solution:
  def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
    subsets: List[List[int]] = []
    permutations: List[int] = []
    def backtrack(index: int):
      subsets.append(permutations[:])
      for i in range(index, len(nums)):
        if i != index and nums[i] == nums[i - 1]:
          continue
        permutations.append(nums[i])
        backtrack(i + 1)
        permutations.pop()
    nums = sorted(nums)
    backtrack(0)
    return subsets